web/src/copy_and_paste.js

```diff
@@ -293,29 +293,138 @@ function get_end_tr_from_endc($endc) {
     return $endc.parents(".selectable_row").first();
 }
 
+function cleanAttribute(attribute) {
+    return attribute ? attribute.replace(/(\n+\s*)+/g, "\n") : "";
+}
+
+function image_to_zulip_markdown(content, node) {
+    if (node.nodeName === "IMG" && node.classList.contains("emoji") && node.hasAttribute("alt")) {
+        // For zulip's custom emojis
+        return node.getAttribute("alt");
+    }
+    const src = node.getAttribute("src") || node.getAttribute("href") || "";
+    const title = cleanAttribute(node.getAttribute("title")) || "";
+    // Using Zulip's link like syntax for images
+    return src ? "[" + title + "](" + src + ")" : node.getAttribute("alt") || "";
+}
+
 export function paste_handler_converter(paste_html) {
-    const turndownService = new TurndownService();
+    // turning off escaping (for now) to remove extra `/`
+    TurndownService.prototype.escape = (string) => string;
+
+    const turndownService = new TurndownService({emDelimiter: "*", codeBlockStyle: "fenced"});
```

Comment:
Hmm, what other settings should we be specifying?  
``` js
    var defaults = {                                                                                    
      rules: rules,                                                                                     
      headingStyle: 'setext',                                                                           
      hr: '* * *',                                                                                      
      bulletListMarker: '*',                                                                            
      codeBlockStyle: 'indented',                                                                       
      fence: '```',                                                                                     
      emDelimiter: '_',                                                                                 
      strongDelimiter: '**',                                                                            
      linkStyle: 'inlined',                                                                             
      linkReferenceStyle: 'full',                                                                       
      br: '  ',                                                                                         
      preformattedCode: false,                                                                          
      ...
```

I would guess that `hr` and `headingStyle` (aka how you make horizontal lines and headings) might need attention?

Reply:
The `hr` default is fine, as we do render that as expected. I updated `headingStyle` to `atx` (`#`s before the heading text, which we do render as headings).

Also had to remove a custom rule removing any formatting for headings (`h1` all the way to `h6`). I traced down the reason we did this from way back in [2017](https://github.com/zulip/zulip/issues/7485#issuecomment-346295026). I'm guessing it was later that we had heading syntax allowed and rendered in the compose box.

Also, I had missed checking for headings while working on this PR so far, since the help center has no mention of them [here](https://zulip.com/help/format-your-message-using-markdown). So it's very likely users won't realise that they can use headings in their messages with standard markdown syntax. Maybe I should create an issue to document this later?

Reply:
> I'm guessing it was later that we had heading syntax allowed and rendered in the compose box.

Yeah, I'm pretty sure we added back heading support sometime after 2017, after doing a bunch of work on the styling to make them not huge.